File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/person/PersonEntity.kt
/*
 * PersonEntity.kt
 *
 * Room @Entity for the `people` table — user-defined people ("Me",
 * "Mom", "Dr. Rao", etc.) that captures can optionally be tagged
 * with. Many-to-many to captures via `capture_people`, mirroring
 * the locations / capture_locations and tags / capture_tags shapes.
 *
 * Seeded with "Me" on first launch (PersonRepository.seedDefaultsIfEmpty).
 */

package app.quickink.mobile.data.person

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "people",
    indices = [
        Index(value = ["user_id", "name"], unique = true),
        Index(value = ["user_id", "position"], name = "idx_people_user_position"),
        Index(value = ["dirty"], name = "idx_people_dirty"),
        Index(value = ["deleted_at"], name = "idx_people_tombstone"),
    ],
)
data class PersonEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "user_id")
    val userId: String,

    @ColumnInfo(name = "name")
    val name: String,

    @ColumnInfo(name = "position", defaultValue = "0")
    val position: Int,

    /**
     * Optional hex color for the person chip (e.g. "#E66943"). NULL →
     * UI falls back to the accent tint.
     */
    @ColumnInfo(name = "color")
    val color: String? = null,

    /**
     * Optional link to a device contact. `lookup_key` is the stable
     * ContactsContract identifier (survives Android contact-id
     * shuffles); `phone`, `email`, `photo_uri` are cached snapshots
     * captured at link time so the row renders consistently without
     * hitting the contacts ContentProvider on every read.
     *
     * Only `phone` and `email` are portable across devices — those
     * travel in the sync payload. `lookup_key` and `photo_uri` are
     * device-local and stay out of the wire shape.
     */
    @ColumnInfo(name = "contact_lookup_key")
    val contactLookupKey: String? = null,

    @ColumnInfo(name = "contact_phone")
    val contactPhone: String? = null,

    @ColumnInfo(name = "contact_email")
    val contactEmail: String? = null,

    @ColumnInfo(name = "contact_photo_uri")
    val contactPhotoUri: String? = null,

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String?,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String?,
)

File: apps/quickink/ios/QuickInk/Data/PersonEntity.swift
/*
 * PersonEntity.swift
 *
 * GRDB record for the `people` table — user-defined people ("Me",
 * "Mom", "Dr. Rao") that captures can optionally be tagged with.
 * Many-to-many to captures via `capture_people`. Mirror of
 * `PersonEntity.kt`.
 *
 * Seeded with a single "Me" row on first launch
 * (PersonRepository.seedDefaultsIfEmpty).
 */

import Foundation
import GRDB

public struct PersonEntity: Codable, FetchableRecord, PersistableRecord, Equatable, Sendable, Identifiable {

    public static let databaseTableName = "people"

    public var id: String
    public var userId: String
    public var name: String
    public var position: Int
    public var color: String?

    /// Optional link to a device contact (CNContact identifier on iOS,
    /// ContactsContract lookup key on Android). Device-local; never
    /// sent to Drive.
    public var contactLookupKey: String?
    public var contactPhone: String?
    public var contactEmail: String?
    /// Cached photo URI (device-local). Never sent to Drive.
    public var contactPhotoUri: String?

    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public init(
        id: String,
        userId: String,
        name: String,
        position: Int,
        color: String? = nil,
        contactLookupKey: String? = nil,
        contactPhone: String? = nil,
        contactEmail: String? = nil,
        contactPhotoUri: String? = nil,
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id               = id
        self.userId           = userId
        self.name             = name
        self.position         = position
        self.color            = color
        self.contactLookupKey = contactLookupKey
        self.contactPhone     = contactPhone
        self.contactEmail     = contactEmail
        self.contactPhotoUri  = contactPhotoUri
        self.driveFileId      = driveFileId
        self.createdAt        = createdAt
        self.updatedAt        = updatedAt
        self.dirty            = dirty
        self.deletedAt        = deletedAt
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case userId           = "user_id"
        case name
        case position
        case color
        case contactLookupKey = "contact_lookup_key"
        case contactPhone     = "contact_phone"
        case contactEmail     = "contact_email"
        case contactPhotoUri  = "contact_photo_uri"
        case driveFileId      = "drive_file_id"
        case createdAt        = "created_at"
        case updatedAt        = "updated_at"
        case dirty
        case deletedAt        = "deleted_at"
    }
}
